packages/@aws-cdk/aws-codepipeline-actions/lib/s3/deploy-action.ts

```diff
@@ -113,12 +113,13 @@ export class S3DeployAction extends Action {
     // the Action Role also needs to read from the Pipeline's bucket
     options.bucket.grantRead(options.role);
 
+    const acl = this.props.accessControl;
     return {
       configuration: {
         BucketName: this.props.bucket.bucketName,
         Extract: this.props.extract === false ? 'false' : 'true',
         ObjectKey: this.props.objectKey,
-        CannedACL: this.props.accessControl,
+        CannedACL: acl ? toKebabCase(acl.toString()) : undefined,
```

Comment:
I may be an uneducated simpleton, but how about this instead?

```ts
        // ...
        CannedACL: this.toCannedAcl(this.props.accessControl),
       // ...

  private toCannedAcl(accessControl: s3.BucketAccessControl | undefined): string | undefined {
    switch (accessControl) {
      case undefined: return undefined;
      case s3.BucketAccessControl.PRIVATE: return 'private';
      case s3.BucketAccessControl.PUBLIC_READ: return 'public-read';
      case s3.BucketAccessControl.PUBLIC_READ_WRITE: return 'public-read-write';
      case s3.BucketAccessControl.AUTHENTICATED_READ: return 'authenticated-read';
      case s3.BucketAccessControl.LOG_DELIVERY_WRITE: return 'log-delivery-write';
      case s3.BucketAccessControl.BUCKET_OWNER_READ: return 'bucket-owner-read';
      case s3.BucketAccessControl.BUCKET_OWNER_FULL_CONTROL: return 'bucket-owner-full-control';
      case s3.BucketAccessControl.AWS_EXEC_READ: return 'aws-exec-read';
    }
  }
```

I understand your code is beautiful and general, but just looking at the regular expression you're using to make this substitution makes my tiny brain hurt, and thus I'm worried about the long-term maintainability of that code.

It also doesn't introduce another public function in `@aws-cdk/core` and keeps the conversation local to the `@awscdk/aws-codepipeline-actions` package, so the 2 modules are less tightly coupled - all good things in my book.

Reply:
as another uneducated simpleton, I'd be glad to make this change, & re-request a review tomorrow. ☺️

Reply:
Actually, if you could make it a stand-alone function (*not* exported from this file), not a method of `S3DeployAction` (it doesn't use any state of the object), that would be even more perfect :).

Reply:
I agree that including a new function toKebabCase is probably overkill. However, we can also avoid the switch/case statements and use a map instead. That makes the map easier to refactor and maintain. Something like:
```
const bucketACLMap = {
    [s3.BucketAccessControl.PRIVATE]: 'private',
    ...etc
}
```

Then:
```
private toCannedAcl(accessControl ...){
 return bucketACLMap[accessControl];
}
```

Reply:
> That makes the map easier to refactor and maintain.

Can you expand what makes the map easier to refactor and maintain?

The map has one huge disadvantage: if there ever is a constant added to `s3. BucketAccessControl`, the compiler will not tell you about it, and you will silently discard it and return `undefined` for that new case. This is not an issue with the `switch` approach.

Reply:
I think, I'll be a passive observer of this conversation, & wait for a decision, before I push any more changes.
cheers

Reply:
True, if there was a missing value... so I'd modify it slightly:

```
private toCannedAcl(accessControl ...){
    return bucketACLMap[accessControl] || accessControl;
}
```

I just hate switch/case statements =-}

I still think the kebabCase stuff is good if we can find a standardized (and tiny) library for it.

Reply:
> ```typescript
> private toCannedAcl(accessControl ...){
>     return bucketACLMap[accessControl] || accessControl;
> }
> ```

That's just silently doing the wrong thing... This is terrible - we will never catch this case until a customer opens us a bug saying "`s3.BucketAccessControl.A_B_C` is resulting in an incorrect value"...

Can you explain why is that form of `toCannedAcl` "easier to refactor and maintain"?

Reply:
Maybe I misspoke about refactor and maintain, it's just less code and, imo, easier to read. If there are functional issues, then fine, but in your switch statement you also don't have a default which means any new BAC added also would silently fail through with an equally bad end user experience "I supplied XXX and CFN is complaining about 'undefined'"

Reply:
In the end, if you feel a switch statement is fine, ok. But considering the fairly well defined rules in turning a BucketAccessControl into a CannedACL (just kebab-case it) I'm not sure why you would't go that route. I doubt AWS is going to introduce a new BAC called "PublicWhatever" and then call the CannedAL "PublicSomethingElse" (which would obviously break with kebab-casing)

Reply:
Folks, I'm confused now.
What if I could make the regular-expression stuff easier to read & manage? 
I'm also open to what @mbonig proposed; to use an external library like [`case`](https://www.npmjs.com/package/case), which is already used in CDK.

Reply:
> but in your switch statement you also don't have a default which means any new BAC added also would silently fail through with an equally bad end user experience "I supplied XXX and CFN is complaining about 'undefined'"

No...? That is absolutely not what would happen...

You would get a compilation error when trying to build the `S3DeployAction` class that your `switch` statement is not exhaustive.